src/inmanta/server/services/compilerservice.py

```diff
@@ -737,39 +744,74 @@ def _compile_merge_key(c: data.Compile) -> Hashable:
+            raise ValueError("The finish_current_compile argument is only supported in single environment mode")
+        if connection is not None and connection.is_in_transaction():
+            raise ValueError("The _process_next_compile_in_queue method should not be called in a transaction context")
 
-    async def _dequeue(self, environment: uuid.UUID) -> None:
-        async with self._global_lock:
-            await self.process_next_compile_in_queue(environment)
+        # acquire connection before lock to ensure lock contention remains short lived
+        async with data.Environment.get_connection(connection) as con:
+            async with self._write_lock_compiling_envs:
+                started_new_compile_run: bool = False
 
-    async def process_next_compile_in_queue(self, environment: uuid.UUID) -> None:
-        if self.is_stopping():
-            return
-        env: Optional[data.Environment] = await data.Environment.get_by_id(environment)
-        if env is None:
-            raise Exception("Can't dequeue compile: environment %s does not exist" % environment)
-        nextrun = await data.Compile.get_next_run(environment)
-        if nextrun and not env.halted:
-            task = self.add_background_task(self._run(nextrun))
-            self._recompiles[environment] = task
-        elif environment in self._recompiles:
-            del self._recompiles[environment]
+                try:
+                    # only inspect state and fetch compile details under the global lock to protect against race conditions
+                    if not self.fully_ready or self.is_stopping():
+                        return
+
+                    compiles: Sequence[data.Compile]
+                    if environment is None:
+                        compiles = await data.Compile.get_next_run_all(connection=con)
+                    else:
+                        c: Optional[data.Compile] = await data.Compile.get_next_run(environment, connection=con)
+                        compiles = [c] if c is not None else []
+
+                    compile: data.Compile
+                    for compile in compiles:
+                        env: Optional[data.Environment] = await data.Environment.get_by_id(compile.environment, connection=con)
+                        if env is None:
+                            raise Exception("Can't queue compile: environment %s does not exist" % compile.environment)
+
+                        # don't execute any compiles in a halted environment
+                        if env.halted:
+                            continue
+
+                        if (
+                            # no compile is running for this environment
+                            compile.environment not in self._compiling_envs
+                            # the previously running compile is finished
+                            or (finish_current_compile and compile.environment == environment)
```

Comment:
I don't get this condition: `compile.environment == environment`?

Reply:
It's not strictly required because there will only ever be at most one compile in the list when `finish_current_compile` is set, due to the method invariants and guards, and it will be a compile for the given environment. In other words the condition will always be met in the current implementation.

So I had the choice between adding the condition or asserting it. I opted for the condition because I felt it's slightly less coupled. But I am not attached to it. If you feel like it is less readable I can go for the assert.

Reply:
Isn't there a big behavioral difference between the assertion and the condition. Namely, the assertion will raise an exception and make us aware about the fact that the invariant was violated somehow. While the condition won't.